Allow variables to be passed in to the template lookup plugin

amenonsen opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
plugins/lookup/template.py

##### ANSIBLE VERSION
```
devel
```

##### SUMMARY

This allows template evaluation to be controlled with inline variable
definitions that augment whatever variables are defined already:

    {{ lookup('template','x.j2',set_variables=dict(x=42)) }}

calfonso commented:
Hi!

Thanks very much for your submission to Ansible. It sincerely means a lot to us that you've taken time to contribute.

Unfortunately, we're not sure if we want this feature in the program, and I don't want this to seem confrontational. Our reasons for this are:

We understand you want to get around declaring the vars in the task with this PR; however, we don't understand why. If you're interested in talking through it and convincing us, we can re-open this PR and review again.

However, we're absolutely always up for discussion. Since this is a really busy project, we don't always see comments on closed tickets, but want to encourage open dialog. You can stop by the development list, and we'd be glad to talk about it - and we might even be persuaded otherwise!

https://groups.google.com/forum/#!forum/ansible-devel
In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

amenonsen committed:
Allow variables to be passed in to template lookup plugin

This allows a single template to be evaluated with different values in
the same task. For example, with a template like 'x:{{a}}', one could do
something like this:

    - foo:
        a: "{{ lookup('template', 'x.j2', template_vars=dict(a=foo[item])) }}"
        b: "{{ lookup('template', 'x.j2', template_vars=dict(a=bar[item])) }}"
      with_items:
        - x
        - y

…and "a" and "b" would expand to different strings based on what we
passed in to the template lookup.

amenonsen commented:
I've pushed an updated version of this PR now. When I originally submitted it, ``templar.set_available_variables()`` did not exist, now there's no need for me to change ``lib/template/__init__.py`` any more.

As explained in the commit message and in a lengthy discussion on IRC, this allows a template to be evaluated different ways within the same task, which is not possible by setting ``vars:`` on a task (in particular, if you want to generate the strings ``"x:1"`` and ``"x:2"`` from the same template ``"x:{{y}}"``, then you can do it by passing in different values of ``y`` to the lookup, but with ``vars:`` you would need to have two different vars ``y1:1`` and ``y2:2`` and two different templates, ``"x:{{y1}}"`` and ``"x:{{y2}}"``.)

I've left the name of the argument as ``vars=`` here, because I can't think of anything better. Maybe ``extra_vars`` could work, but anything else seems like a gratuitous change, e.g., ``variables=``. Suggestions?

bcoca commented on docs/docsite/rst/playbooks_lookups.rst:
```diff
@@ -556,6 +556,9 @@ Here are some examples::
 
          - debug: msg="{{ lookup('template', './some_template.j2') }} is a value from evaluation of this template"
 
+         # passing in variables during evaluation
```
add 'since version 2.4'

bcoca commented on lib/ansible/plugins/lookup/template.py:
```diff
@@ -36,6 +36,7 @@ class LookupModule(LookupBase):
     def run(self, terms, variables, **kwargs):
 
         convert_data_p = kwargs.get('convert_data', True)
+        extra_vars = kwargs.get('vars', {})
```
from core meeting, the vote is in and we would like the new option to be named `template_vars`
